File: 141.linked-list-cycle.cpp
/*
 * @lc app=leetcode id=141 lang=cpp
 *
 * [141] Linked List Cycle
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_set<ListNode*> list;
        while(head!=NULL){
            if (list.count(head))  return true;
            else list.insert(head);
            head=head->next;
        }
        return false;
    }
};
/* 方法一 哈希表法，将所有的节点指针存入哈希表中，一旦出现重复，这说明
出现了重复，则说明出现了链表环 返回true，如遍历链表中所有指针后没有出现
重复 则返回false


方法二 快慢指针法，设置两个指针，一个一次跳一步，一个一次跳两步，那么如果
存在环形链表，则最后快的指针
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head==NULL||head->next==NULL) return false;
        ListNode* slow=head;
        ListNode* fast=head;
        while(slow!=head){
            if (fast->next==NULL||fast==NULL) return false;
            slow=slow->next;
            fast=fast->next->next;
        }
        return ture;
    }
};

File: 141.linked-list-cycle.py
#
# @lc app=leetcode id=141 lang=python
#
# [141] Linked List Cycle
#
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head==None:
            return False
        hs={head}
        head=head.next
        while head:
            if head in hs:
                return True
            else:
                hs.add(head)
                head=head.next
        return False 
'''此为python的容器法，使用一个容器来保留唯一的指针，如果出现重复则说明出现
环

方法二：快慢指针法
class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head==None or head.next==None:
            return False
        f=head.next
        s=head
        while f!=s:
            if f==None or f.next==None:
                return False
            else:
                s=s.next
                f=f.next.next
        return True 
思路与c++的思路一致
